src/daft-scan/src/scan_task_iters/split_parquet_files_by_rowgroup.rs

```diff
@@ -0,0 +1,465 @@
+                            MaybeSplitScanTaskByParquetRowgroups::NoSplit(_) => None,
+                            MaybeSplitScanTaskByParquetRowgroups::Split(split) => {
+                                Some((idx, split))
+                            }
+                        },
+                    );
+
+                    // Perform a split of ScanTasks: note that this might trigger expensive I/O to retrieve Parquet metadata
+                    let split_scan_tasks = split_scan_tasks_by_parquet_metadata(
+                        scan_tasks_to_split.collect_vec().as_slice(),
+                        self.config,
+                    );
+                    let mut split_scan_tasks = match split_scan_tasks {
+                        Err(e) => {
+                            self.state =
+                                SplitParquetFilesByRowGroupsState::ConstructingWindow(vec![]);
+                            return Some(Err(e));
+                        }
+                        Ok(split_scan_tasks) => split_scan_tasks,
+                    };
+
+                    // Zip the split_scan_tasks back into the iterator by substituting entries that have been split using their index
+                    let mut window_after_split = VecDeque::<ScanTaskRef>::new();
+                    for (idx, maybe_split_scan_task) in finalized_window.into_iter().enumerate() {
+                        if let Some(splits) = split_scan_tasks.remove(&idx) {
+                            window_after_split.extend(splits);
+                        } else {
+                            window_after_split.push_back(maybe_split_scan_task.unwrap());
+                        }
+                    }
+                    self.state =
+                        SplitParquetFilesByRowGroupsState::EmittingSplitWindow(window_after_split);
+                    continue;
+                }
+                // Emitting split window:
+                //
+                // Exhaust the window after splitting, and when completely exhausted, re-enter the ConstructingWindow state.
+                SplitParquetFilesByRowGroupsState::EmittingSplitWindow(window_after_split) => {
+                    if let Some(item) = window_after_split.pop_front() {
+                        return Some(Ok(item));
+                    } else {
+                        self.state =
+                            SplitParquetFilesByRowGroupsState::ConstructingWindow(Vec::new());
+                        continue;
+                    }
+                }
+            }
+        }
+    }
+}
+
+#[must_use]
+pub(crate) fn split_all_files_by_rowgroup<'a>(
+    scan_tasks: BoxScanTaskIter<'a>,
+    config: &'a DaftExecutionConfig,
+) -> BoxScanTaskIter<'a> {
+    Box::new(SplitParquetFilesByRowGroups {
+        iter: scan_tasks,
+        state: SplitParquetFilesByRowGroupsState::ConstructingWindow(Vec::new()),
+        config,
+    }) as BoxScanTaskIter
+}
```

Comment:
I'm also considering maybe making this 2 different passes instead of mixing Parquet metadata retrieval and file splitting into the same pass.

1. `window_batched_parquet_metadata(scan_tasks) -> BoxScanTaskIter`: hydrates ScanTasks by retrieving Parquet metadata in batches, pending some predicates (e.g. max window size, size of the file etc)
2. `split_by_parquet_rowgroup(scan_tasks) -> BoxScanTaskIter`: performs splits of individual ScanTasks if the required metadata is available

LMK if this is desirable instead? I think this could be a nicer design and easier to reason about/extend. For example, for Iceberg:

1. `window_batched_parquet_metadata` can be a no-op because we can detect that Iceberg has populated the required metadata
2. `split_by_parquet_rowgroup` will use that Iceberg provided metadata to split the file (no windowing logic required here)